src/main/java/org/logstash/plugins/outputs/s3/GzipUtil.java

```diff
@@ -0,0 +1,96 @@
+package org.logstash.plugins.outputs.s3;
+
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Ruby Gzip interfaces (such as GzipReader) have limitations to access and manage the ZIP files.
+ * Where Java has a deep control over ZIP files.
+ * This class introduces GZIP related operations.
+ */
+public class GzipUtil {
+
+    private final static int FILE_READ_SIZE = 1024;
+
+    private final static String CORRUPTED_FILE_ERROR = "Unexpected end of ZLIB input stream";
+
+    private final static Logger logger = LogManager.getLogger(GzipUtil.class);
+
+    /**
+     * Compresses a given non-compressed source file into compressed target file.
+     * @param sourcePath a path where source is located
+     * @param targetPath a path where target can be placed
+     * Note: raises an IO exception if any break happens.
+     */
+    public static void compress(String sourcePath, String targetPath) {
+        try {
+            GzipUtil.compress(Paths.get(sourcePath), Paths.get(targetPath));
+        } catch (IOException exception) {
+            logger.error("Error occurred while compressing the file, error={}", exception.getMessage());
+        }
+    }
+
+    /**
+     * Decompresses and recovers corrupted GZIP file into target file.
+     * @param sourcePath a path where corrupted GZIP is located
+     * @param targetPath a path where target can be placed
+     * Note: raises an IO exception other than GZIP dead blocks.
+     */
+    public static void recover(String sourcePath, String targetPath) {
+        try {
+            GzipUtil.decompress(Paths.get(sourcePath), Paths.get(targetPath));
+        } catch (IOException exception) {
+            // raise an exception if expected exception is not end of ZLIB related.
+            if (CORRUPTED_FILE_ERROR.equals(exception.getMessage()) == false) {
+                logger.error("Error occurred while compressing the file, error={}", exception.getMessage());
```

Comment:
Well, so far the only case when GZIP might be corrupted is `no tail` case where stream is left opened when LS is crashed. In case if other cases happen, currently, LS will leave file as it is, users will be able to find it under temporary path. No any delete action will be proceeded. And also, most importantly, we don't damage/crash LS and let it continue.
Let me know your thought if we can improve the situation.

Reply:
Unless, I'm missing something, I don't think LS will leave the file as it is. With this change, the error will be logged, but processing will continue as normal - uploading the corrupt file to S3, and deleting the file from local storage.

Reply:
File path changes when corrupted file is recovered (`file_path = self.recover(file_path)`) and new temp file will be returned (with `create_from_existing_file`). If LS does not recover the file its size is still zero (in temporary file) and it does not remote the file ([source](https://github.com/mashhurs/logstash-output-s3/blob/60ba59e392f2e515c2d2ad2fd6d16dded1636569/lib/logstash/outputs/s3.rb#L404))

Reply:
If the file size is 0, the temporary file _and_ the original "corrupt" file will be deleted - the process of deleting a "temporary file" also [deletes the folder](https://github.com/logstash-plugins/logstash-output-s3/blob/main/lib/logstash/outputs/s3/temporary_file.rb#L52) it is a part of.

Reply:
Oh it is because of temp folder... yeah agree that temp folder will be removed. Thanks for letting know, reproducing such case is not easy. How do you think if we replace following logic 

```
// s3.rb:404L
        if temp_file.size > 0
        @logger.debug? && @logger.debug("Recovering from crash and uploading", :path => temp_file.path)
        @crash_uploader.upload_async(temp_file,
                                     :on_complete => method(:clean_temporary_file),
                                     :upload_options => upload_options)
      end
      if @encoding != "gzip" && temp_file.size != 0
        clean_temporary_file(temp_file)
      end
```

Reply:
But I don't see the logic where deleting the folder... Can you plz let me know?
For any unsuccessful conditions, In `clean_temporary_file(temp_file)` new file for recovery (not original corrupted ZIP) will be passed and removed which is okay.